dynamiq/nodes/agents/orchestrators/graph.py

```diff
@@ -0,0 +1,319 @@
+            end_state = State(id=END, description="Final state")
+            self._state_by_id[END] = end_state
+            self.states.append(end_state)
+
+    @property
+    def to_dict_exclude_params(self):
+        return super().to_dict_exclude_params | {"manager": True, "states": True}
+
+    def to_dict(self, **kwargs) -> dict:
+        """Converts the instance to a dictionary.
+
+        Returns:
+            dict: A dictionary representation of the instance.
+        """
+        data = super().to_dict(**kwargs)
+        data["manager"] = self.manager.to_dict(**kwargs)
+        data["states"] = [state.to_dict(**kwargs) for state in self.states]
+        return data
+
+    def add_state(self, state_id: str, tasks: list[Node | Callable]) -> None:
+        """
+        Adds state with specified tasks to the graph.
+
+        Args:
+            state_id (str): Id of the state.
+            tasks (list[Agent | Callable]): List of tasks that have to be executed when running this state.
+
+        Raises:
+            ValueError: If state with specified name already exists.
+        """
+        if state_id in self._state_by_id:
+            raise ValueError(f"Error: State with name {state_id} already exists.")
+
+        filtered_tasks = []
+
+        has_agent = False
+        for task in tasks:
+            if isinstance(task, Node):
+                if isinstance(task, Agent):
+                    has_agent = True
+                filtered_tasks.append(task)
+            elif isinstance(task, Callable):
+                filtered_tasks.append(function_tool(task)())
+            else:
+                raise OrchestratorError("Error: Task must be either a Node or a Callable.")
+
+        state = State(id=state_id, name=state_id, manager=self.manager if has_agent else None, tasks=filtered_tasks)
+        self.states.append(state)
+        self._state_by_id[state.id] = state
+
+    def add_edge(self, source: str, destination: str) -> None:
+        """
+        Adds edge to the graph.
+
+        Args:
+            source (str): Name of source state.
+            destination (str): Name of destinations state.
+
+        Raises:
+            ValueError: If state with specified name is not present.
+        """
+        self.validate_states([source, destination])
+        self._state_by_id[source].next_states = [destination]
+
+    def validate_states(self, names: list[str]) -> None:
+        """
+        Check if the provided state names is valid (exists in the list of valid states).
+
+        Args:
+            names (list[str]): names of states to validate.
+
+        Raises:
+            ValueError: If state with specified name is not present.
+        """
+        for state_name in names:
+            if state_name not in self._state_by_id:
+                raise ValueError(f"State with name {state_name} is not present")
+
+    def add_conditional_edge(self, source: str, destinations: list[str], condition: Callable | Python) -> None:
```

Comment:
Why here we have `destinations: list[str]`, and in `def add_edge` `destination: str`?

Reply:
`add_edge`  accepts source and destination to create edge. This is used in case we want to go to destination right after source is finished. In `add_conditional_edge` providing destinations is not obligatory for functionality, but critical to ability to draw this graph. In case of conditional edge next state is determined by passed function or Python Node.

Reply:
I mean, why does conditional edge support multiple destinations and edge just one?
If there is some specific - please describe it in docstring.